### ISSUE TITLE
Question: should `observer` always apply pure?

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx-react#160]
Currently `observer` tries do a shallow comparison on next available properties. However if there is a data structure (object) passed into the props which is not an observable object, mobx-react assumes it might have been deeply mutated (similar to the assumption react makes) and re-renders to not take any changes. 

Sometimes this is confusing for people, as they might expect shallow / pure comparison *always*. Should this be changed? Is it worth the potential risk of accidentally optimizing too much renders away?

cc: @Strate @andykog @ConneXNL

**Comment by mweststrate** [comment]
Ok, asking the question is answering it I think.

**Comment by andykog** [comment]
@mweststrate, 
1. This behaviour can be configurable like in redux connect.
2. Or `mobx-react` can export `pureRendering` checker (or `dirtyRendering` depending on what to choose as a default behaviour, so one can use it like:
```js
import observer, { pureRendering } from 'mobx-react';
class extends Component {
  shouldComponentUpdate(...args) {
    return pureRendering(...args);
  }
  render() {}
}
```
I would be comfortable with `pure` by default

**Comment by mweststrate** [comment]
Released as part of 4.0.0
